src/compiler/checker.ts

```diff
@@ -39752,6 +39752,28 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     }
                     if (leftOk && rightOk) {
                         checkAssignmentOperator(resultType);
+                        switch (operator) {
+                            case SyntaxKind.LessThanLessThanToken:
+                            case SyntaxKind.LessThanLessThanEqualsToken:
+                            case SyntaxKind.GreaterThanGreaterThanToken:
+                            case SyntaxKind.GreaterThanGreaterThanEqualsToken:
+                            case SyntaxKind.GreaterThanGreaterThanGreaterThanToken:
+                            case SyntaxKind.GreaterThanGreaterThanGreaterThanEqualsToken:
+                                const rhsEval = evaluate(right);
+                                if (typeof rhsEval.value === "number" && Math.abs(rhsEval.value) >= 32) {

```

Comment:
Should this convert the value to an integer or something?

Reply:
The `Math.abs` also seems unnecessary - any value below 0 seems wrong and `1 << -1 == 1 << 31`, `1 << -2 == 1 << 30` etc. so `Math.abs` "reverses" it.

Interestingly, `1 << -0.9 == 1 << 0` so I guess you want `Math.trunc` (or `| 0`) rather than `Math.floor`.

Reply:
> any value below 0 seems wrong

It's a useful shorthand. Similar to negative slice indices in python. Not doing an `abs` on the evaluated result here, either - just using it to measure the magnitude of the number to issue the error or not.

> Should this convert the value to an integer or something?

Eh. Don't really need to. Fractional parts just get dropped anyway. Issuing an error on *any* fractional literal in a shift position would probably be fine (`0.5` is the same as `0` for this operator after all), but I don't think it's particularly related to this range error.